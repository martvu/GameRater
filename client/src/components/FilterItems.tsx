import { useState } from 'react';
import { Plus, Minus } from 'lucide-react';
import { Checkbox } from './ui/checkbox';
import { Button } from './ui/button';
import { useRecoilState, useSetRecoilState } from 'recoil';
import {
  pageState,
  selectedGenresState,
  selectedPlatformsState,
} from '@/state/atoms.ts';
import { Genre, Platform } from '@/gql/graphql';
import { Label } from './ui/label';

interface FilterItemsProps {
  filters: Platform[] | Genre[];
  filterType: 'platforms' | 'genres';
}

export default function FilterItems({ filters, filterType }: FilterItemsProps) {
  const [numItemsToShow, setNumItemsToShow] = useState(10);
  const [selectedPlatforms, setSelectedPlatforms] = useRecoilState(
    selectedPlatformsState
  );
  const [selectedGenres, setSelectedGenres] =
    useRecoilState(selectedGenresState);
  const setCurrentPage = useSetRecoilState(pageState);
  const showMore = () => {
    setNumItemsToShow(numItemsToShow + 10);
  };

  const handleCheckboxChange = (filterId: number, isChecked: boolean) => {
    if (filterType === 'platforms') {
      setSelectedPlatforms(
        isChecked
          ? [...selectedPlatforms, filterId]
          : selectedPlatforms.filter(id => id !== filterId)
      );
    } else {
      setSelectedGenres(
        isChecked
          ? [...selectedGenres, filterId]
          : selectedGenres.filter(id => id !== filterId)
      );
    }
    setCurrentPage(1); // Reset page to 1 when a filter is selected
  };

  const showLess = () => setNumItemsToShow(10);
  return (
    <div>
      <h2 className="my-2 text-left font-semibold capitalize text-foreground">
        {filterType}
      </h2>
      <div className="ml-1 flex flex-col">
        {filters?.slice(0, numItemsToShow).map(item => (
          <div className="my-1 flex items-center space-x-2" key={item?.name}>
            <Checkbox
              id={`filter-item-${item?.name}`}
              onCheckedChange={checked => {
                if (typeof item?.id === 'number')
                  handleCheckboxChange(item.id, checked.valueOf() as boolean);
              }}
              checked={
                filterType === 'platforms'
                  ? selectedPlatforms.includes(item?.id ?? 0)
                  : selectedGenres.includes(item?.id ?? 0)
              }
            />
            <Label
              htmlFor={`filter-item-${item?.name}`}
              className={`text-left font-normal tracking-tight ${
                (filterType === 'platforms' &&
                  selectedPlatforms.includes(item?.id ?? 0)) ||
                (filterType === 'genres' &&
                  selectedGenres.includes(item?.id ?? 0))
                  ? 'text-foreground'
                  : 'text-muted-foreground'
              }`}
            >
              {item?.name}
              {/* {item?.gamesCount && ` (${item?.gamesCount})`} */}
            </Label>
          </div>
        ))}
      </div>
      {filters && numItemsToShow < filters.length ? (
        <Button
          variant={'text'}
          onClick={showMore}
          className="flex w-36 items-center pl-0 text-green-400"
        >
          <Plus className="inline-block h-6" />
          <span>Show More</span>
        </Button>
      ) : numItemsToShow > 10 && filters.length > 10 ? (
        <Button
          variant={'text'}
          onClick={showLess}
          className="flex w-36 items-center pl-0 text-red-400"
        >
          <Minus className="inline-block h-6" />
          <span>Show Less</span>
        </Button>
      ) : (
        <div></div>
      )}
    </div>
  );
}
