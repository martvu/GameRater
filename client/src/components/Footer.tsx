export default function Footer() {
  return (
    <footer className="mt-5 h-16 w-full border-t text-muted-foreground">
      <div className="container mx-auto px-6 pt-5 text-left text-sm">
        Copyright 2023 GameRater
      </div>
    </footer>
  );
}
