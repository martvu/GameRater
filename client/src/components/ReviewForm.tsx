import { zodResolver } from '@hookform/resolvers/zod';
import * as z from 'zod';
import { Button } from '@/components/ui/button';

import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { useForm } from 'react-hook-form';
import { Textarea } from './ui/textarea';
import { StarRating } from './Rating';
import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select';
import { useQuery, gql, useMutation } from '@apollo/client';
import { useParams } from 'react-router-dom';

const GET_GAME = gql`
  query GetGamePlatforms($id: ID!) {
    getGame(ID: $id) {
      name
      platforms {
        id
        name
      }
    }
  }
`;

const CREATE_REVIEW = gql`
  mutation CreateReview($reviewInput: ReviewInput!) {
    createReview(reviewInput: $reviewInput) {
      author
      title
      content
      rating
      platform
    }
  }
`;

type GameDetailParams = {
  id: string;
};

const formSchema = z.object({
  title: z.string().min(2, {
    message: 'Username must be at least 2 characters.',
  }),
  content: z.string().max(1000).min(4),
  rating: z
    .number()
    .min(1, {
      message: 'Please select a rating.',
    })
    .max(5),
  platform: z.string().min(1, {
    message: 'Please select a platform.',
  }),
});

export function ReviewForm() {
  const { id } = useParams<GameDetailParams>();
  const { loading, error, data } = useQuery(GET_GAME, {
    variables: { id: id },
  });
  const [createReview] = useMutation(CREATE_REVIEW);
  // 1. Define your form.
  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      title: '',
      content: '',
      rating: 0,
      platform: '',
    },
  });

  // 2. Define a submit handler.
  function onSubmit(values: z.infer<typeof formSchema>) {
    //Handle submit
    try {
      createReview({
        variables: {
          reviewInput: {
            title: values.title,
            content: values.content,
            rating: values.rating,
            platform: values.platform,
            author: 'me',
            gameID: id,
          },
        },
        onCompleted: () => {
          console.log('Review created');
        },
      });
    } catch (error) {
      console.log('Could not create review');
    }
    //Reset form
    form.reset();
    //Refresh page
    window.location.reload();
    console.log(values);
  }

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :</p>;

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
        <FormField
          control={form.control}
          name="title"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Review Title</FormLabel>
              <FormControl>
                <Input placeholder="Title of your review" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="platform"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Platform</FormLabel>
              <Select onValueChange={field.onChange} defaultValue={field.value}>
                <FormControl>
                  <SelectTrigger>
                    <SelectValue placeholder="Choose Platform" />
                  </SelectTrigger>
                </FormControl>
                <SelectContent>
                  <SelectGroup>
                    {data.getGame.platforms.map(
                      (platform: { name: string }) => (
                        <SelectItem key={platform.name} value={platform.name}>
                          {platform.name}
                        </SelectItem>
                      )
                    )}
                  </SelectGroup>
                </SelectContent>
              </Select>

              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="content"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Review Content</FormLabel>
              <FormControl>
                <Textarea
                  placeholder="Tell us what you think about this game"
                  className="resize-none"
                  {...field}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="rating"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Your Rating</FormLabel>
              <FormControl>
                <StarRating
                  disabled={false}
                  rating={field.value}
                  onChange={field.onChange}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormDescription>
          Your review will be public and shared with other users.
        </FormDescription>
        <Button type="submit">Submit Review</Button>
      </form>
    </Form>
  );
}

//
