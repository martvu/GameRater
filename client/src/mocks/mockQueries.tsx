import {
  ADD_FAVORITES,
  CREATE_REVIEW,
  REMOVE_FAVORITES,
  SIGN_IN_OR_CREATE_USER,
} from '@/lib/mutations';
import {
  GET_GAMES,
  GET_GAME,
  GET_FILTERS,
  GET_GAME_PLATFORMS,
  GET_SEARCH_SUGGESTIONS,
} from '@/lib/queries';

const getGamesMock = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: undefined,
      userId: '',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            user_rating: 4.5,
          },
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getGamesMockedKeyword = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: 'mockedKeyword',
      userId: '',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            user_rating: 4.5,
          },
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getGamesSignedInMock = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: 'mockedKeyword',
      userId: '1',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            platforms: [1, 2],
            genres: [3, 4],
          },
          // Add more game objects as needed
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getFiltersMock = {
  request: {
    query: GET_FILTERS,
    variables: {},
  },
  result: {
    data: {
      getGenres: [
        { id: 1, name: 'Genre 1' },
        { id: 2, name: 'Genre 2' },
      ],
      getPlatforms: [
        { id: 1, name: 'Platform 1' },
        { id: 3, name: 'Platform 3' },
        { id: 4, name: 'PC (Microsoft Windows)' },
        { id: 2, name: 'Nintendo Switch' },
        { id: 5, name: 'PlayStation 5' },
        { id: 6, name: 'Xbox Series X|S' },
      ],
    },
  },
};

const getGamePlatformsMock = {
  request: {
    query: GET_GAME_PLATFORMS,
    variables: { id: '1' },
  },
  result: {
    data: {
      getGame: {
        name: 'Example Game',
        platforms: [
          {
            id: 3,
            name: 'Platform 3',
          },
          {
            id: 4,
            name: 'PC (Microsoft Windows)',
          },
        ],
      },
    },
  },
};

const signInMock = {
  request: {
    query: SIGN_IN_OR_CREATE_USER,
    variables: {
      userInput: {
        username: 'testuser',
      },
    },
  },
  result: {
    data: {
      signInOrCreateUser: {
        _id: '1',
        username: 'testuser',
        favorites: [],
        reviews: [],
      },
    },
  },
};

const getGameMock = {
  request: {
    query: GET_GAME,
    variables: {
      id: '1',
      limit: 5,
      offset: 0,
      username: '',
    },
  },
  result: {
    data: {
      getGame: {
        _id: '1',
        name: 'Example Game',
        summary: 'This is an example game summary.',
        imageId: 'image1',
        first_release_date: 1620000000,
        aggregatedRating: 85,
        user_rating: 4.5,
        platforms: [{ name: 'PC' }],
        genres: [{ name: 'Adventure' }],
        reviews: {
          count: 2,
          reviews: [
            {
              _id: 'review1',
              user: 'user1',
              title: 'Great Game',
              content: 'This is a great game because...',
              rating: 5,
              platform: 'PC',
              gameID: '1',
            },
            {
              _id: 'review2',
              user: 'user2',
              title: 'Good Game',
              content: 'This is a good game because...',
              rating: 4,
              platform: 'PC',
              gameID: '1',
            },
          ],
          userHasReviewed: false,
        },
      },
    },
  },
};

const getGameMock2 = {
  request: {
    query: GET_GAME,
    variables: {
      id: '1',
      limit: 5,
      offset: 0,
      username: 'testUser',
    },
  },
  result: {
    data: {
      getGame: {
        _id: '1',
        name: 'Example Game',
        summary: 'This is an example game summary.',
        imageId: 'image1',
        first_release_date: 1620000000,
        aggregatedRating: 85,
        user_rating: 4.5,
        platforms: [{ name: 'PC' }],
        genres: [{ name: 'Adventure' }],
        reviews: {
          count: 2,
          reviews: [
            {
              _id: 'review1',
              user: 'user1',
              title: 'Great Game',
              content: 'This is a great game because...',
              rating: 5,
              platform: 'PC',
              gameID: '1',
            },
            {
              _id: 'review2',
              user: 'user2',
              title: 'Good Game',
              content: 'This is a good game because...',
              rating: 4,
              platform: 'PC',
              gameID: '1',
            },
          ],
          userHasReviewed: false,
        },
      },
    },
  },
};
const addFavoritesMock = {
  request: {
    query: ADD_FAVORITES,
    variables: { username: 'testUser', gameID: '1' },
  },
  result: {
    data: {
      addFavorites: {
        _id: '1',
        username: 'testUser',
        favorites: [{ _id: '1' }],
      },
    },
  },
};

const removeFavoritesMock = {
  request: {
    query: REMOVE_FAVORITES,
    variables: { username: 'testUser', gameID: '1' },
  },
  result: {
    data: {
      removeFavorites: {
        _id: '1',
        username: 'testUser',
        favorites: [],
      },
    },
  },
};

const createReviewMock = {
  request: {
    query: CREATE_REVIEW,
    variables: {
      reviewInput: {
        title: 'Great Game',
        content: 'This is a great game because...',
        rating: 5,
        platform: 'PC (Microsoft Windows)',
        user: 'testUser',
        gameID: '1',
      },
    },
  },
  result: {
    data: {
      createReview: {
        _id: 'review1',
        user: 'testUser',
        title: 'Great Game',
        content: 'This is a great game because...',
        rating: 5,
        platform: 'PC (Microsoft Windows)',
        gameID: '1',
      },
    },
  },
};

const searchbarSuggestionsMock = {
  request: {
    query: GET_SEARCH_SUGGESTIONS,
    variables: {
      query: 'example',
    },
  },
  result: {
    data: {
      getSearchSuggestions: [
        {
          _id: '1',
          name: 'Example Game',
          cover_image_id: 'cover1',
        },
        {
          _id: '2',
          name: 'Example Game 2',
          cover_image_id: 'cover2',
        },
      ],
    },
  },
};

const allMocks = [
  createReviewMock,
  getGamePlatformsMock,
  getGamesMockedKeyword,
  addFavoritesMock,
  removeFavoritesMock,
  getGamesMock,
  getFiltersMock,
  signInMock,
  getGamesSignedInMock,
  getGameMock,
  getGameMock2,
  searchbarSuggestionsMock,
  // Add more mocks as needed for other queries
];

export {
  allMocks,
  getGamesMock,
  getGamesSignedInMock,
  getFiltersMock,
  signInMock,
};
