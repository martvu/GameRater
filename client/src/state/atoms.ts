import { atom } from 'recoil';
import { GameSortInput, User } from '../gql/graphql';

const user = localStorage.getItem('user');
const defaultUser = {
  _id: '',
  username: '',
  favorites: [],
  reviews: [],
};

export const userState = atom<User>({
  key: 'userState',
  default: user ? JSON.parse(user) : defaultUser,
});

export const selectedPlatformsState = atom({
  key: 'selectedPlatformsState',
  default: [] as number[],
});

export const selectedGenresState = atom({
  key: 'selectedGenresState',
  default: [] as number[],
});

const currentPage = localStorage.getItem('currentPage');
export const pageState = atom<number>({
  key: 'pageState',
  default: currentPage ? parseInt(currentPage, 10) : 1,
});

const sortBy = localStorage.getItem('selectedSortState');

export const defaultSortBy = {
  field: 'first_release_date',
  order: 'desc',
};

export const sortByState = atom<GameSortInput>({
  key: 'sortByState',
  default: sortBy ? JSON.parse(sortBy) : defaultSortBy,
});

