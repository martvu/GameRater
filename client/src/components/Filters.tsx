import { useState } from 'react';
import { ChevronLeft, ListFilter } from 'lucide-react';
import FilterItems from './FilterItems';
import { Button } from './ui/button';
import { cn } from '@/lib/utils';
import { GET_FILTERS } from '@/lib/queries';
import { useQuery } from '@apollo/client';
import { Genre, Platform } from '@/gql/graphql';
import { useRecoilValue } from 'recoil';
import {
  genresListState,
  platformsListState,
  selectedGenresState,
  selectedPlatformsState,
} from '@/state/atoms';
import { ScrollArea } from './ui/scroll-area';

/**
 * Filters component
 * Displayed on large screens (non-mobile)
 */
export default function Filters() {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const genreList = useRecoilValue(genresListState);
  const platformList = useRecoilValue(platformsListState);
  const selectedGenres = useRecoilValue(selectedGenresState);
  const selectedPlatforms = useRecoilValue(selectedPlatformsState);

  const toggleCollapse = () => {
    setIsCollapsed(!isCollapsed);
  };
  const { data } = useQuery(GET_FILTERS);

  const preferredOrder = [
    'Nintendo Switch',
    'PC (Microsoft Windows)',
    'PlayStation 5',
    'Xbox Series X|S',
  ];

  // Custom sorting function to sort platforms by preferred order
  const sortByPreference = (a: Platform, b: Platform) => {
    const aIndex = preferredOrder.indexOf(a.name as string);
    const bIndex = preferredOrder.indexOf(b.name as string);

    if (aIndex !== -1 && bIndex !== -1) {
      return aIndex - bIndex; // Both are preferred, sort by preferred order
    } else if (aIndex !== -1) {
      return -1; // Only a is preferred
    } else if (bIndex !== -1) {
      return 1; // Only b is preferred
    }
    return 0; // Neither is preferred, maintain original order
  };

  // Sort the platforms
  // Create a copy of the platforms array before sorting
  const platformsCopy: Platform[] = [...(data?.getPlatforms ?? [])];
  const genresCopy: Genre[] = [...(data?.getGenres ?? [])];
  // Now sort the copied array
  const sortedPlatforms: Platform[] = platformsCopy.sort(sortByPreference);

  // Filter only platforms in platformList (when search limits platforms) or all platforms if platformList is empty
  const filteredPlatforms = sortedPlatforms.filter(platform => {
    if (platformList.length === 0) return true;
    return (
      platformList.includes(platform.id as number) ||
      selectedPlatforms.includes(platform.id as number)
    );
  });

  // Filter only genres in genreList or all genres if genreList is empty
  const filteredGenres = genresCopy.filter(genre => {
    if (genreList.length === 0) return true;
    return (
      genreList.includes(genre.id as number) ||
      selectedGenres.includes(genre.id as number)
    );
  });

  // Apply transition to the filter menu
  const filterContainerClasses = cn(
    'flex h-full flex-col items-start overflow-hidden text-left transition-all duration-300 ease-in-out',
    isCollapsed ? 'w-0' : 'w-[240px]' // Toggle between full width and collapsed state
  );

  return (
    <>
      <section className={filterContainerClasses}>
        <ScrollArea className="h-full pb-4 pr-2">
          <section className="pb-4 pl-5 md:pb-32">
            <h1
              className={`mt-4 text-left text-xl font-bold tracking-wider text-foreground ${
                isCollapsed ? 'hidden' : ''
              }`}
            >
              Filter By
            </h1>
            {!isCollapsed && data && (
              <>
                <FilterItems
                  filters={filteredPlatforms}
                  filterType="platforms"
                />
                <FilterItems filters={filteredGenres} filterType="genres" />
              </>
            )}
          </section>
        </ScrollArea>
      </section>
      <Button
        className="fixed bottom-8 left-8 z-50 m-1 hidden md:flex"
        variant="secondary"
        onClick={toggleCollapse}
        aria-label={isCollapsed ? 'show filters' : 'hide filters'}
      >
        {isCollapsed ? <ListFilter /> : <ChevronLeft size={18} />}
        {!isCollapsed && <span className="ml-2">Hide filters</span>}
      </Button>
    </>
  );
}
