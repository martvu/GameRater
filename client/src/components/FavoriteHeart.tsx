import { Heart } from 'lucide-react';
import { Button } from './ui/button';
import { useEffect, useState } from 'react';
import { useRecoilState } from 'recoil';
import { userState } from '@/state/atoms';
import { Game } from '@/gql/graphql';
import { useMutation } from '@apollo/client';
import { ADD_FAVORITES, REMOVE_FAVORITES } from '@/lib/mutations';
import { Dialog, DialogContent } from './ui/dialog';
import SignInAlertModal from './SignInAlertModal';

interface FavoriteHeartProps {
  game: Game;
  variant?: 'secondary' | 'ghost' | 'text' | 'outline';
}

/**
 * FavoriteHeart component
 * @param {Game} game - The game object
 * @param {string} variant - The variant of the button (default ghost)
 */
export default function FavoriteHeart({ game, variant }: FavoriteHeartProps) {
  const [addFavorites] = useMutation(ADD_FAVORITES);
  const [removeFavorite] = useMutation(REMOVE_FAVORITES);
  const [user, setUser] = useRecoilState(userState);
  const [isUpdating, setIsUpdating] = useState(false);
  const [isDialogOpen, setIsDialogOpen] = useState(false);

  const toggleFavorite = async (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    event.preventDefault();
    event.stopPropagation();
    // If the user is not signed in, open the sign in alert modal
    if (!user.username) {
      setIsDialogOpen(true);
      return;
    }
    if (isUpdating) return;
    setIsUpdating(true);
    if (!user.username) {
      setIsUpdating(false);
      return;
    }
    const isFavorite = user.favorites?.some(
      favoriteGame => favoriteGame?._id === game._id
    );
    const updatedFavorites = isFavorite
      ? user.favorites?.filter(
          favoriteGame => favoriteGame && favoriteGame._id !== game._id
        )
      : [...(user.favorites || []), game];

    // Optimistically update the UI
    setUser({ ...user, favorites: updatedFavorites });

    try {
      await (isFavorite ? removeFavorite : addFavorites)({
        variables: {
          username: user.username as string,
          gameID: game._id as string,
        },
      });
    } catch (error) {
      console.log('Could not update favorites', error);
      setUser(user);
    } finally {
      setIsUpdating(false);
    }
  };

  const handleSignInSuccess = () => {
    setIsDialogOpen(false);
  };

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user));
  }, [user]);

  const isFavorite = user.favorites?.some(
    favoriteGame => favoriteGame?._id === game._id
  );

  return (
    <>
      <Button
        aria-label="Favorite game button"
        data-testid="favorite-btn"
        onClick={toggleFavorite}
        className="z-10"
        variant={variant || 'ghost'}
        size="round"
      >
        <Heart
          data-testid="heart-icon"
          className={isFavorite ? 'fill-red-600 text-red-600' : 'none'}
        />
      </Button>
      <Dialog open={isDialogOpen} onOpenChange={setIsDialogOpen}>
        <DialogContent>
          <SignInAlertModal
            message={'Please sign in to favorite this game!'}
            onSignInSuccess={handleSignInSuccess}
          />
        </DialogContent>
      </Dialog>
    </>
  );
}
