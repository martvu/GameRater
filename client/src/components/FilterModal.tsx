import { Button } from '@/components/ui/button';
import {
  Dialog,
  DialogClose,
  DialogContent,
  DialogFooter,
  DialogTrigger,
} from '@/components/ui/dialog';
import Filters from './Filters';

/**
 * FilterModal component
 * Displayed on small screens (mobile devices)
 */
export default function FilterModal() {
  return (
    <Dialog>
      <DialogTrigger asChild>
        <Button variant="outline" className="w-[150px]">
          Filters
        </Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-[425px]">
        <Filters />
        <DialogFooter>
          <DialogClose asChild>
            <Button type="submit">Apply Filter</Button>
          </DialogClose>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
