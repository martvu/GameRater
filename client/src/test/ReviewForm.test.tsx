import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MockedProvider } from '@apollo/client/testing';
import { RecoilRoot } from 'recoil';
import { ReviewForm } from '@/components/ReviewForm';
import { allMocks } from '@/mocks/mockQueries';
import { vi } from 'vitest';
import { userState } from '@/state/atoms';

class MockPointerEvent extends Event {
  button: number;
  ctrlKey: boolean;
  pointerType: string;

  constructor(type: string, props: PointerEventInit) {
    super(type, props);
    this.button = props.button || 0;
    this.ctrlKey = props.ctrlKey || false;
    this.pointerType = props.pointerType || 'mouse';
  }
}
//eslint-disable-next-line @typescript-eslint/no-explicit-any
window.PointerEvent = MockPointerEvent as any;
window.HTMLElement.prototype.scrollIntoView = vi.fn();
window.HTMLElement.prototype.releasePointerCapture = vi.fn();
window.HTMLElement.prototype.hasPointerCapture = vi.fn();

vi.mock('react-router-dom', async () => {
  const actualModule =
    await vi.importActual<typeof import('react-router-dom')>(
      'react-router-dom'
    );

  return {
    ...actualModule,
    useParams: vi.fn(() => ({ id: '1' })),
  };
});

const mockUser = {
  _id: '1',
  username: 'testUser',
  favorites: [],
  reviews: [],
};

const renderReviewForm = () => {
  return render(
    <MockedProvider mocks={allMocks} addTypename={false}>
      <RecoilRoot
        initializeState={({ set }) => {
          set(userState, mockUser);
        }}
      >
        <ReviewForm />
      </RecoilRoot>
    </MockedProvider>
  );
};
describe('ReviewForm', () => {
  it('matches the snapshot', async () => {
    const { asFragment } = renderReviewForm();

    await waitFor(() => {
      expect(screen.getByText('Submit Review')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', async () => {
    renderReviewForm();

    await waitFor(() => {
      const titleInput = screen.getByTestId('review-title-input');
      expect(titleInput).toBeInTheDocument();
      expect(screen.getByText('Submit Review')).toBeInTheDocument();
    });
  });

  it('can be submitted', async () => {
    renderReviewForm();
    // Fill out the form
    const titleInput = await screen.findByTestId('review-title-input');
    await userEvent.type(titleInput, 'Great Game');
    const submitButton = await screen.findByText('Submit Review');
    fireEvent.click(submitButton);
    await waitFor(() => {
      expect(
        screen.getByText('Content must be at least 4 characters.')
      ).toBeInTheDocument();
      expect(screen.getByText('Please select a platform.')).toBeInTheDocument();
      expect(screen.getByText('Please select a rating.')).toBeInTheDocument();
    });
    const platformSelect = await screen.findByTestId('review-platform-select');
    await userEvent.click(platformSelect);
    await userEvent.click(screen.getAllByText('PC (Microsoft Windows)')[1]);
    const contentInput = await screen.findByTestId('review-content-input');
    await userEvent.type(contentInput, 'This is a great game because...');
    const ratingInput = await screen.findAllByTestId('review-star');
    await userEvent.click(ratingInput[4]);
    // Submit the form
    fireEvent.click(submitButton);
  });
});
