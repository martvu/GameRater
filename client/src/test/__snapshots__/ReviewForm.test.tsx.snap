// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReviewForm > matches the snapshot 1`] = `
<DocumentFragment>
  <form
    class="space-y-8"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Review Title
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="review-title-input"
        id=":r0:-form-item"
        name="title"
        placeholder="Title of your review"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Platform
      </label>
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r2:"
        aria-describedby=":r1:-form-item-description"
        aria-expanded="false"
        aria-invalid="false"
        class="flex h-10 w-full items-center justify-between rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        data-placeholder=""
        data-state="closed"
        data-testid="review-platform-select"
        dir="ltr"
        id=":r1:-form-item"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Choose Platform
        </span>
        <svg
          aria-hidden="true"
          class="h-4 w-4 opacity-50"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <select
        aria-hidden="true"
        style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        tabindex="-1"
      >
        <option
          value="Platform 3"
        >
          Platform 3
        </option>
        <option
          value="PC (Microsoft Windows)"
        >
          PC (Microsoft Windows)
        </option>
      </select>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r3:-form-item"
      >
        Review Content
      </label>
      <textarea
        aria-describedby=":r3:-form-item-description"
        aria-invalid="false"
        class="flex min-h-[80px] w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 resize-none"
        data-testid="review-content-input"
        id=":r3:-form-item"
        name="content"
        placeholder="Tell us what you think about this game"
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r4:-form-item"
      >
        Your Rating
      </label>
      <div
        class="flex items-center font-light"
      >
        <button
          class="text-muted-foreground"
          data-testid="review-star"
          type="button"
        >
          <span
            class="text-2xl font-light"
          >
            ☆  
          </span>
        </button>
        <button
          class="text-muted-foreground"
          data-testid="review-star"
          type="button"
        >
          <span
            class="text-2xl font-light"
          >
            ☆  
          </span>
        </button>
        <button
          class="text-muted-foreground"
          data-testid="review-star"
          type="button"
        >
          <span
            class="text-2xl font-light"
          >
            ☆  
          </span>
        </button>
        <button
          class="text-muted-foreground"
          data-testid="review-star"
          type="button"
        >
          <span
            class="text-2xl font-light"
          >
            ☆  
          </span>
        </button>
        <button
          class="text-muted-foreground"
          data-testid="review-star"
          type="button"
        >
          <span
            class="text-2xl font-light"
          >
            ☆  
          </span>
        </button>
        <span
          class="ml-2 text-muted-foreground"
        >
          0/5
        </span>
      </div>
    </div>
    <p
      class="text-sm text-muted-foreground"
      id="undefined-form-item-description"
    >
      Your review will be public and shared with other users.
    </p>
    <button
      class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 active:bg-primary/70 h-10 px-4 py-2"
      data-testid="review-submit-btn"
      type="submit"
    >
      Submit Review
    </button>
  </form>
</DocumentFragment>
`;
